### PR TITLE
+ Adds a new Tabbed Settings page

### PR DESCRIPTION
**Pull request by shadyvb**
- Adds option to disable minification alltogether
- Adds option to disable minification based on specific conditions ( islogged, is_admin, query var present )
- Adds option to exclude resources, line-delimited, based on string search not matching, means it matches any part of the url
- Note: Disabling minification does not disable serving minified files, so cached pages does not break.

### CODE REVIEW

**Comment by lkraav** [comment]
Trying this functionality out, because s2member plugin doing some funky stuff with its enqueueing and needs to be excluded.

First comment: commit messages should be better.

http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

**Comment by lkraav** [comment]
Second, I think you have some whitespace issues in a7f751e9f19e6415e3f161cfb68b52055ea625d9 visible in at least the constants section in the very start of the commit. You should not be changing constant order or upstream code whitespace format.

**Comment by westonruter** [comment]
Re: commit messages, yes they should ideally be composed according to that [note](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html). Although the prefixing of each line with `+` or `*` is kinda helpful, this information can also be obtained with a log stat; so we try to have the commit summary be in the imperative form, like:

> Add a new Tabbed Settings page

or 

> Expose default option for exclusion

Not a huge deal though.

**Comment by lkraav** [comment]
Functionality: what I'd like to see for URL exclusion is also a filter implementation, so code-based overrides become possible. Then admin UI should also use the filter to do its thing.

Starting with just the filter would probably be the minimal viable product that has a chance of getting merged quick.

**Comment by lkraav** [comment]
Meh, I guess the filter is already done https://github.com/x-team/wp-dependency-minification/blob/master/dependency-minification.php#L707

**Comment by westonruter** [comment]
Yes, re: filter, see the forums: http://wordpress.org/support/topic/woocommerce-not-supported?replies=2#post-4555036

**Comment by shadyvb** [comment]
@lkraav 
- The note just makes sense, thanks for that.
- Regarding white spaces, not sure if you noticed, but i added a -slightly longer- [constant](https://github.com/x-team/wp-dependency-minification/commit/a7f751e9f19e6415e3f161cfb68b52055ea625d9#L2R22), so i had to realign all lines, as per Wordpress coding standards stated in [Contributing.md](https://github.com/x-team/wp-dependency-minification/blob/master/contributing.md)

File: dependency-minification.php [hunk 1]
```diff
@@ -49,7 +54,21 @@ static function setup() {
 			||
 			in_array( $GLOBALS['pagenow'], array( 'wp-login.php', 'wp-register.php' ) )
 		);
-		if ( $is_frontend ) {
+		$disabled = (
+			! empty( self::$options['disabled_on_conditions']['all'] )
```
Comment (westonruter) [inline dependency-minification.php]: @shadyvb `Notice: Undefined index: all`

File: dependency-minification.php [hunk 2]
```diff
@@ -49,7 +54,21 @@ static function setup() {
...
+			|| ( !empty( self::$options['disabled_on_conditions']['loggedin'] ) && is_user_logged_in() )
```
Comment (westonruter) [inline dependency-minification.php]: @shadyvb `Notice: Undefined index: loggedin` etc

File: dependency-minification.php [hunk 3]
```diff
@@ -447,6 +497,69 @@ static function admin_page() {
...
+										<input type="checkbox" name="options[disabled_on_conditions][queryvar][enabled]" id="options[disabled_on_conditions][queryvar][enabled]" <?php echo checked( self::$options['disabled_on_conditions']['queryvar']['enabled'] ) ?> value="1">
+										<?php esc_html_e( 'Query Variable', 'depmin' ) ?>
+										<input type="text" name="options[disabled_on_conditions][queryvar][value]" id="options[disabled_on_conditions][queryvar][value]" value="<?php echo esc_html( self::$options['disabled_on_conditions']['queryvar']['value'] ) ?>">
+									</label>
+								</td>
+							</tr>
+							<tr>
+								<th><?php esc_html_e( 'Exclude resources', 'depmin' ) ?>
+									<small><?php esc_html_e( 'Add URLs of resources to exclude from minification, one resource per line. Note that you can just add a script name, or the last portion of the URL so it matches.', 'depmin' ) ?></small></th>
+								<td><textarea name="options[exclude_dependencies]" id="options[exclude_dependencies]" rows="10" class="widefat"><?php echo esc_html( join( "\n", self::$options['exclude_dependencies'] ) ) ?></textarea></td>
```
Comment (westonruter) [inline dependency-minification.php]: @shadyvb `Warning:  join(): Invalid arguments passed in`

File: dependency-minification.php [hunk 4]
```diff
@@ -479,6 +529,69 @@ static function admin_page() {
...
+									<label for="options[disabled_on_conditions][loggedin]">
+										<input type="checkbox" name="options[disabled_on_conditions][loggedin]" id="options[disabled_on_conditions][loggedin]" <?php echo checked( self::$options['disabled_on_conditions']['loggedin'] ) ?> value="1">
+										<?php esc_html_e( 'Logged in Users', 'depmin' ) ?>
+									</label>
+									<label for="options[disabled_on_conditions][admin]">
+										<input type="checkbox" name="options[disabled_on_conditions][admin]" id="options[disabled_on_conditions][admin]" <?php echo checked( self::$options['disabled_on_conditions']['admin'] ) ?> value="1">
+										<?php esc_html_e( 'Administrators', 'depmin' ) ?>
+									</label>
+									<label for="options[disabled_on_conditions][queryvar][enabled]">
+										<input type="checkbox" name="options[disabled_on_conditions][queryvar][enabled]" id="options[disabled_on_conditions][queryvar][enabled]" <?php echo checked( self::$options['disabled_on_conditions']['queryvar']['enabled'] ) ?> value="1">
```
Comment (nash-ye) [inline dependency-minification.php]: You don't have to call "echo" for checked() function.

File: dependency-minification.php [hunk 5]
```diff
@@ -479,6 +529,69 @@ static function admin_page() {
 					</table>
 				<?php endif; ?>
 			</form>
+			</div>
+			<?php
+			$nonce = wp_create_nonce( self::AJAX_OPTIONS_ACTION );
+			?>
+			<div class="nav-tab-content" id="tab-content-settings">
+				<form action="<?php echo admin_url( 'admin-ajax.php' ) ?>" method="post" class="form-table options">
```
Comment (nash-ye) [inline dependency-minification.php]: You may need to use esc_url() here.

File: dependency-minification.php [hunk 6]
```diff
@@ -290,8 +337,11 @@ static function admin_page() {
 		<div class="wrap">
 			<div class="icon32" id="icon-tools"><br></div>
```
Comment (nash-ye) [inline dependency-minification.php]: Why you don't use `screen_icon()` function
http://codex.wordpress.org/Function_Reference/screen_icon

File: dependency-minification.php [hunk 7]
```diff
@@ -270,6 +290,33 @@ static function admin_ajax_handler() {
 	}
 
 	/**
+	 * @action wp_ajax_dependency_minification_options
+	 */
+	static function admin_ajax_options_handler() {
+		if ( ! current_user_can( self::$options['admin_page_capability'] ) ) {
+			wp_die( __( 'You are not allowed to do that.', 'depmin' ) );
+		}
+		if ( ! wp_verify_nonce( $_REQUEST['_wpnonce'], self::AJAX_OPTIONS_ACTION ) ) {
```
Comment (nash-ye) [inline dependency-minification.php]: I think you should check by `$_POST['_wpnonce']` ?!

File: dependency-minification.php [hunk 8]
```diff
@@ -270,6 +290,33 @@ static function admin_ajax_handler() {
...
+			wp_die( __( 'Nonce check failed. Try reloading the previous page.', 'depmin' ) );
+		}
+		$updated_count = 0;
+		if ( ! empty( $_REQUEST['options'] ) ) {
```
Comment (nash-ye) [inline dependency-minification.php]: hmm.. What's wrong with `$_POST` ?!

**Comment by shadyvb** [comment]
Did some fixes to comply with WPCS in code, not only related to PR, only remaining issues are mostly related to indentation.

UPDATE: @westonruter I just figured out that i shouldn't have fixed CS issues on this PR, it'd make it quite hard for code review. Sorry!

File: dependency-minification.php [hunk 9]
```diff
@@ -186,7 +204,8 @@ static function admin_notices() {
 			?>
 			<div class="error">
 				<p><?php
-				echo sprintf(
+				echo esc_html(
```
Comment (nash-ye) [inline dependency-minification.php]: I didn't test this yet.. but how you use `esc_html` with some HTML tags in the string? Can you check this.

Comment (shadyvb) [inline dependency-minification.php]: Good catch! Fixed in https://github.com/shadyvb/wp-dependency-minification/commit/a18558f2df1f43ef92f412a1b5b562724b376cb7

**Comment by westonruter** [comment]
This has been open too long, sorry @shadyvb. Now that it is in `develop` we can work toward the next release. One thing I need to do is add Travis CI.

**Comment by westonruter** [comment]
Opened PR https://github.com/x-team/wp-dependency-minification/pull/44 to track additional fixes for PHP_CodeSniffer, and for the 1.0 release.

**Pull request merged**